src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -1258,5 +1284,65 @@ public override void Process()
             }
         }
     }
+
+    public sealed class HashEstimator : IEstimator<HashTransform>
+    {
+        public const int NumBitsMin = 1;
+        public const int NumBitsLim = 32;
+
+        public static class Defaults
+        {
+            public const int HashBits = NumBitsLim - 1;
+            public const uint Seed = 314489979;
+            public const bool Ordered = false;
+            public const int InvertHash = 0;
+        }
+
+        private readonly IHost _host;
+        private readonly HashTransform.ColumnInfo[] _columns;
+
+        public HashEstimator(IHostEnvironment env, string name, string source = null,
+            int hashBits = Defaults.HashBits, int invertHash = Defaults.InvertHash)
+            : this(env, new HashTransform.ColumnInfo(name, source ?? name, hashBits: hashBits, invertHash: invertHash))
+        {
+        }
+
+        public HashEstimator(IHostEnvironment env, params HashTransform.ColumnInfo[] columns)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(nameof(HashEstimator));
+            _columns = columns.ToArray();
+            foreach (var col in _columns)
+            {
+                if (col.InvertHash < -1)
+                    throw _host.ExceptParam(nameof(columns), "Value too small, must be -1 or larger");
```

Comment:
>ExceptParam [](start = 32, length = 11)

why is it not part of ctor for `ColumnInfo`?

Reply:
I had same thoughts, mainly because we don't have IExceptionContext in ColumnInfo, but yeah

---
In reply to: [218908992](https://github.com/dotnet/machinelearning/pull/944#discussion_r218908992) [](ancestors = 218908992)

Reply:
Yea, I would be ok with losing context info here for the sake of localizing the validation

---
In reply to: [218909546](https://github.com/dotnet/machinelearning/pull/944#discussion_r218909546) [](ancestors = 218909546,218908992)